kernel/src/process/process/mod.rs

```diff
@@ -343,254 +338,274 @@ impl Process {
+        process_group_mut: &mut MutexGuard<Weak<ProcessGroup>>,
+        session_table_mut: &mut MutexGuard<BTreeMap<Sid, Arc<Session>>>,
+        group_table_mut: &mut MutexGuard<BTreeMap<Pgid, Arc<ProcessGroup>>>,
+    ) -> Sid {
+        let (session, process_group) = Session::new_pair(self.clone());
+        let sid = session.sid();
 
-        // Creates a new session
-        let new_session = Session::new(new_group.clone());
-        let mut new_group_inner = new_group.inner.lock();
-        new_group_inner.session = Arc::downgrade(&new_session);
-        new_session.inner.lock().leader = Some(self.clone());
-        session_table_mut.insert(new_session.sid(), new_session.clone());
+        **process_group_mut = Arc::downgrade(&process_group);
 
-        // Removes the process from session.
-        let mut session_inner = session.inner.lock();
-        session_inner.remove_process(self);
+        // Insert the new session and the new process group to the global table.
+        session_table_mut.insert(session.sid(), session);
+        group_table_mut.insert(process_group.pgid(), process_group);
 
-        Ok(new_session)
+        sid
     }
 
-    /// Moves the process to other process group.
+    /// Moves the process to another process group.
     ///
-    ///  * If the group already exists, the process and the group should belong to the same session.
-    ///  * If the group does not exist, this method creates a new group for the process and move the
-    ///    process to the group. The group is added to the session of the process.
+    /// If `pgid` is equal to the process ID, a new process group with the given PGID will be
+    /// created (if it does not already exist). Then, the process will be moved to the process
+    /// group with the given PGID, if the process group exists and belongs to the same session as
+    /// the given process.
     ///
-    /// This method may return `EPERM` in following cases:
-    ///  * The process is session leader;
-    ///  * The group already exists, but the group does not belong to the same session as the process;
-    ///  * The group does not exist, but `pgid` is not equal to `pid` of the process.
-    pub fn to_other_group(self: &Arc<Self>, pgid: Pgid) -> Result<()> {
-        // if the process already belongs to the process group
-        if self.pgid() == pgid {
-            return Ok(());
-        }
+    /// # Errors
+    ///
+    /// This method will return `EPERM` in following cases:
+    ///  * `session_referer` is `Some(_)`, but the process is not in the same session;
+    ///  * The process is a session leader, but the given PGID is not the process's PID/PGID;
+    ///  * The process group already exists, but the group does not belong to the same session;
+    ///  * The process group does not exist, but `pgid` is not equal to the process ID.
+    pub fn to_other_group(
+        self: &Arc<Self>,
+        pgid: Pgid,
+        session_referer: Option<&Self>,
```

Comment:
The `session_referer` can only refer to the current process, correct? In that case, I recommend renaming it to `current`. Additionally, it would be prudent to add a debug check to ensure that if `self` is not the current process, `session_referer` should contain some value.

Reply:
The code here has a few more problems.
https://github.com/asterinas/asterinas/blob/3edacc870843e09c327cb1e97683b1f3731d4f59/kernel/src/process/process/mod.rs#L499
So we cannot take the process group table lock after the process table lock.

Maybe you can check the diff here: https://github.com/asterinas/asterinas/compare/3edacc870843e09c327cb1e97683b1f3731d4f59..b9b0b7e1e5a3e1573e02443a7390bac292ea5db9

Reply:
But the latest code converts it to `self.or_child_to_group(pid, pgid)`. This may not be perfect, but at least it is more correct?

Reply:
Yes, the latest code seems to not have the reverse lock order problem and looks fine to me. But the method `or_child_to_group` maybe better renamed as `move_process_to_group`? You should already documented that the process can only be `self` or a `child` process.